### ISSUE TITLE
let-binding has unit value fix breaks code

### ISSUE THREAD

**Issue by Eagle941** [rust-lang/rust-clippy#10758]
### Summary

This is the code:
```rust
pub struct AnyId<T: ?Sized> {
   metadata: <T as Pointee>::Metadata,
}

pub struct AnyStore {}

impl AnyStore {
   pub fn push<T: 'static>(value: T) -> AnyId<T> {
      let metadata = std::ptr::metadata(&value);
      AnyId {
	   metadata,
      }
   }
}
```

I have got this warning:
```
warning: this let-binding has unit value
  --> dyn_utils\src\any_store.rs:97:3
   |
97 |         let metadata = std::ptr::metadata(&value);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: omit the `let` binding: `std::ptr::metadata(&value);`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
   = note: `#[warn(clippy::let_unit_value)]` on by default
```

I tried this command:

```
cargo clippy --fix --lib -p dyn_utils
```

This happened:
```
The following errors were reported:
error[E0425]: cannot find value `metadata` in this scope
   --> dyn_utils\src\any_store.rs:111:4
    |
111 |             metadata,
    |             ^^^^^^^^ not found in this scope
    |
help: consider importing one of these items
    |
1   | use core::ptr::metadata;
    |
1   | use std::fs::metadata;
    |
1   | use std::ptr::metadata;
    |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: this let-binding has unit value
  --> dyn_utils\src\any_store.rs:97:3
   |
97 |         let metadata = std::ptr::metadata(&value);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: omit the `let` binding: `std::ptr::metadata(&value);`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
   = note: `#[warn(clippy::let_unit_value)]` on by default

warning: `dyn_utils` (lib) generated 1 warning (run `cargo clippy --fix --lib -p dyn_utils` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.48s
```

### Lint Name

let_unit_value

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

### Version

```text
rustc 1.71.0-nightly (f9a6b7158 2023-05-05)
binary: rustc
commit-hash: f9a6b71580cd53dd4491d9bb6400f7ee841d9c22
commit-date: 2023-05-05
host: x86_64-pc-windows-msvc
release: 1.71.0-nightly
LLVM version: 16.0.2
```

**Comment by Centri3** [comment]
@blyxyas, [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=752911d663ad4a61542ca00d9942ed39) (minimized)

This one seems tricky. The docs of `std::ptr::Pointee` state that `Sized` pointers are thin and the metadata is `()`, so while this seems to be correct this should still not be linted, if possible

PS, the `?Sized` bound in `AnyId` is likely meaningless here since `push` isn't `?Sized`, which means that it'll always be `()`.

**Comment by Centri3** [comment]
Also, this probably shouldn't suggest removing the binding if it's used later

**Comment by matthiaskrgr** [comment]
Another example where clippy would remove `v` while it is later used
````rust
// check-pass

#![feature(unboxed_closures)]

trait Lt<'a> {
    type T;
}
impl<'a> Lt<'a> for () {
    type T = ();
}

fn main() {
    let v: <() as Lt<'_>>::T = (); // => ();
    let f: &mut dyn FnMut<(_,), Output = ()> = &mut |_: <() as Lt<'_>>::T| {};
    f(v);
}
````

**Comment by ijijn** [comment]
Indeed! Here's another example, admittedly simplified to the point of absurdity...

```rust
fn process<T>(_input: T) {
    // ...
}

fn main() {
    let data = ();
    process(data);
}
```

**Comment by sgdxbc** [comment]
+1 on @ijijn case. The fixing should come with inlining but it does not.
